File: Heap_Sort/Heap_Sort.php
<?php
  
  /*
    
    Heap Sort is a comparison-based sorting technique performed on a Heap data structure
        
  */
 
function MaxHeapify(&$data, $heapSize, $index) {
    $left = ($index + 1) * 2 - 1; // left = 2*i + 1 
    $right = ($index + 1) * 2; // right = 2*i + 2 
    $largest = 0;

    // To heapify a subtree rooted with node i which is 
    // an index in arr[]. n is size of heap 
     if ($left < $heapSize && $data[$left] > $data[$index])  // If left child is larger than root 
         $largest = $left;
     else
         $largest = $index;
 
     if ($right < $heapSize && $data[$right] > $data[$largest]) // If right child is larger than largest so far 
         $largest = $right;
  
     if ($largest != $index)  // If largest is not root 
     {
         $temp = $data[$index];
         $data[$index] = $data[$largest];
         $data[$largest] = $temp;
 
         //Recursively heapify the affected sub-tree 
         MaxHeapify($data, $heapSize, $largest);
      }
}
 
function HeapSort(&$data, $count) {
    $heapSize = $count;
 
    // Build heap (rearrange array) 
    for ($p = ($heapSize - 1) / 2; $p >= 0; $p--)
        MaxHeapify($data, $heapSize, $p);
 
    // One by one extract an element from heap
    for ($i = $count - 1; $i > 0; $i--)
    {
        $temp = $data[$i];
        $data[$i] = $data[0];
        $data[0] = $temp;
 
        $heapSize--;
        MaxHeapify($data, $heapSize, 0);
    }
}
 
// example of driver function
$array = array(20 , 43 , 65 , 88 , 11 , 33 , 56 , 74);
HeapSort($array , 8);
print_r($array);
 
 // output
   /*Array
   (
    [0] => 11
    [1] => 20
    [2] => 33
    [3] => 65
    [4] => 43
    [5] => 56
    [6] => 74
    [7] => 88
    )
    
*/

?>

File: Heap_Sort/heap_sort.go

/*	Heap Sort is a comparison-based sorting method performed on a Heap data structure.
    Here we build a max heap first to later sort the elements in the heap in ascending order.
*/
   

package main 
import "fmt" 


func max_heapify(arr []int, n int, i int) { 

    largest := i // Initialize largest as root 
    l := 2 * i + 1 // left child
    r := 2 * i + 2 // right child
  
    // If left child is larger than root 
	if l < n && arr[l] > arr[largest]{
        largest = l
		}
  
    // If right child is larger than largest so far 
    if r < n && arr[r] > arr[largest]{	
        largest = r
		}
  
    // If largest is not root 
    if largest != i{ 
        arr[i], arr[largest] = arr[largest], arr[i] //swapping
  
        // Recursively heapify the affected sub-tree 
        max_heapify(arr, n, largest) 
    } 
} 


func heapSort(arr []int, n int ) { 

    // Build heap (rearrange array) 
    for  i := n / 2 - 1; i >= 0; i-- { 
        max_heapify(arr, n, i)
    }
  
    // One by one extract an element from heap 
    for  i := n - 1; i >= 0; i-- { 
        // Move current root to end 
        arr[0], arr[i]=arr[i],arr[0];  //swapping
  
        // call heapify on the reduced heap 
        max_heapify(arr, i, 0); 
    } 
}
 

func printArray(arr []int, no int) {  
   
    //printing the sorted array
    for i := 0; i <= 10; i++ {
        fmt.Printf("\n%d",arr[i])
    }
    
}


func main() {
	var n int
	fmt.Println("Enter the number of elements :") 
	fmt.Scan(&n) //accepting the number of elements from the user
	arr := make([]int, n) //creating an array of n elements
	
	for i := 0; i < n; i++ {
      fmt.Scan(&arr[i])  //accepting input elements
    }
	fmt.Println(arr)
    
    fmt.Println("Sorted array is") 
	heapSort(arr, n)
    printArray(arr, n) 
}

/* OUTPUT:

Enter the number of elements : 7
Enter the elements :
0
12
11
13
5
6
7

Original Array: [0 12 11 13 5 6 7]

Sorted array is
0 5 6 7 11 12 13 

*/